cpp/include/rapidsmpf/shuffler/chunk.hpp

```diff
@@ -25,145 +27,335 @@ namespace rapidsmpf::shuffler::detail {
+     *
+     * @param chunk_id The ID of the chunk.
+     * @param part_id The ID of the partition.
+     * @param packed_data The packed data.
+     * @param event The CUDA event.
+     * @param stream The CUDA stream.
+     * @param br The buffer resource.
+     * @return The ChunkBatch.
+     */
+    static Chunk from_packed_data(
+        ChunkID chunk_id,
+        PartID part_id,
+        PackedData&& packed_data,
+        std::shared_ptr<Buffer::Event> event,
+        rmm::cuda_stream_view stream,
+        BufferResource* br
     );
 
     /**
-     * @brief Returns an unpacked (deserialized) chunk.
+     * @brief Create a single-message ChunkBatch for a finished partition (control
+     * message).
      *
-     * @warning This copies the data and shouldn't be used in performance critical code.
+     * @param chunk_id The ID of the chunk.
+     * @param part_id The ID of the partition.
+     * @param expected_num_chunks The expected number of chunks.
+     * @return The ChunkBatch.
+     */
+    static Chunk from_finished_partition(
+        ChunkID chunk_id, PartID part_id, size_t expected_num_chunks
+    );
+
+    /**
+     * @brief Create a ChunkBatch by deserializing a metadata message.
+     *
+     * @param msg The metadata message received from another rank.
+     * @param validate Whether to validate the metadata buffer.
+     * @return The ChunkBatch.
      *
-     * @param stream CUDA stream used for device memory operations and kernel launches.
-     * @returns A `cudf::table` that represents the chunk data.
+     * @throws std::runtime_error if the metadata buffer does not follow the expected
+     * format and `validate` is true.
      */
-    [[nodiscard]] std::unique_ptr<cudf::table> unpack(rmm::cuda_stream_view stream) const;
+    static Chunk deserialize(std::vector<uint8_t> const& msg, bool validate = true);
 
     /**
-     * @brief Returns a description of this instance.
+     * @brief Validate if a deserialized buffer follows the Chunk format.
+     *
+     * @param serialized_buf The deserialized buffer to validate.
+     * @return True if the deserialized buffer follows the Chunk format, false
+     * otherwise.
+     */
+    static bool validate_format(std::vector<uint8_t> const& serialized_buf);
+
+    /**
+     * @brief Whether the chunk is ready for consumption.
+     *
+     * @return True if the chunk is ready, false otherwise.
+     * @note ChunkBatch is ready if it has no data or if the data is ready. data_ buffer
+     * could be set later, so we need to check if it is non-null.
+     */
+    [[nodiscard]] inline bool is_ready() const {
+        // data_offsets_[-1] contains the size of the data buffer. If it is 0, the chunk
+        // has no data messages, so it is ready. Else, the chunk is ready if the data
+        // buffer is non-null and the data buffer is ready.
+        return !data_offsets_.empty()
+               && (data_offsets_[n_messages() - 1] == 0 || (data_ && data_->is_ready()));
```

Comment:
*nit*: if `data_offsets_` is not empty, under what circumstances can `data_` be `nullptr`?

i.e. why is this not:

```
return !data_offsets_.empty() && (data_.size() == 0 || data_->is_ready());
```

Reply:
When we receive messages, we first receive concat metadata buffer, and then we deserialize it to form a valid `Chunk` that knows data buffer size, but it's not allocated. https://github.com/nirandaperera/rapidsmpf/blob/multi-packed-data-chunk/cpp/src/shuffler/shuffler.cpp#L206
Allocation happens later, 
https://github.com/nirandaperera/rapidsmpf/blob/multi-packed-data-chunk/cpp/src/shuffler/shuffler.cpp#L237-L238